File: 63-UniquePathsII/dp.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int rows = obstacleGrid.length;
        if (rows == 0) return 0;
        int cols = obstacleGrid[0].length;
        
        int[][] dp = new int[rows][cols];
        boolean reach = true;
        for (int i = 0; i < rows; i++) {
            if (obstacleGrid[i][0] == 1) reach = false;
            dp[i][0] = reach ? 1 : 0;
        }
        reach = true;
        for (int j = 0; j < cols; j++) {
            if (obstacleGrid[0][j] == 1) reach = false;
            dp[0][j] = reach ? 1 : 0;
        }

        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                if (obstacleGrid[i][j] == 1) dp[i][j] = 0;
                else dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[rows-1][cols-1];
    }
}

File: 931-MinFallingPathSum/dp.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& A) {
        int rows = A.size();
        if (!rows) return 0;
        int cols = A[0].size();
        int res = INT_MAX;
        vector<int> dp(cols);
        for (int i = 0; i < cols; ++i) dp[i] = A[0][i];
        for (int i = 1; i < rows; ++i) {
            vector<int> tmp(dp.begin(), dp.end());
            for (int j = 0; j < cols; ++j) {
                int temp = dp[j];
                if (j > 0) temp = min(temp, dp[j - 1]);
                if (j < cols - 1) temp = min(temp, dp[j + 1]);
                tmp[j] = temp + A[i][j];
            }
            for (int i = 0; i < cols; ++i) dp[i] = tmp[i];
        }
        for (int i = 0; i < cols; ++i) res = min(res, dp[i]);
        return res;
    }
};